s4e-web/cypress/integration/map/map-layers.spec.ts

```diff
@@ -1,141 +1,290 @@
+
+      cy.server();
+      cy.route({
+        method: 'GET',
+        url: `${geoserverUrl}?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities`,
+        headers: {
+          'Content-type': 'text/xml'
+        },
+        status: 404,
+        response:  { errors: 'CannotGET /...' }
+      })
+        .as('getCapabilities');
+
+      MapLayers
+        .openManagementModal()
+        .fillFormCorrectly(label, geoserverUrl)
+        .errorShouldBeDispalyed();
+        GeneralModal
+        .closeModal();
+    });
+
+    it('handle 500 error', function () {
+      const label = 'Test';
+      const geoserverUrl = '/test-wms';
+
+      cy.server();
+      cy.route({
+        method: 'GET',
+        url: `${geoserverUrl}?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities`,
+        headers: {
+          'Content-type': 'text/xml'
+        },
+        status: 500,
+        response:  { errors: 'CannotGET /...' }
+      })
+        .as('getCapabilities');
+      MapLayers
+        .openManagementModal()
+        .fillFormCorrectly(label, geoserverUrl)
+        .errorShouldBeDispalyed();
+        GeneralModal
+        .closeModal();
+    });
+
+    it('handle 503 error', function () {
+      const label = 'Test';
+      const geoserverUrl = '/test-wms';
+
+      cy.server();
+      cy.route({
+        method: 'GET',
+        url: `${geoserverUrl}?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities`,
+        headers: {
+          'Content-type': 'text/xml'
+        },
+        status: 503,
+        response:  { errors: 'CannotGET /...' }
+      })
+        .as('getCapabilities');
+        
+      MapLayers
+        .openManagementModal()
+        .fillFormCorrectly(label, geoserverUrl)
+        .errorShouldBeDispalyed();
+        GeneralModal
+        .closeModal();
+    });
+
+    it('handle response does not have the correct formats ', function () {
+      const label = 'Test';
+      const geoserverUrl = '/test-wms';
 
-  it('should display url errors', function () {
-    const label = 'Test';
-    const url = [
-      'https://test-page.pl/?SERVICE=unknown',
-      'REQUEST=unknown',
-      'LAYERS=unkn?own',
-      'STYLES=unkn?own',
-      'FORMAT=image/unknown',
-      'TRANSPARENT=none',
-      'VERSION=x.y.z',
-      'HEIGHT=none',
-      'WIDTH=none',
-      'CRS=unknown',
-      'BBOX=unknown,unknown,unknown,unknown'
-    ].join('&');
-
-    Layers
-      .openManagementModal()
-      .addNew(label, url)
-      .errorsCountShouldBe(6)
-      .changeContextTo(GeneralModal)
-      .closeAndChangeContext(Layers);
+      cy.server();
+      cy.route({
+        method: 'GET',
+        url: `${geoserverUrl}?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities`,
+        headers: {
+          'Content-type': 'text/xml'
+        },
+        status: 200,
+        response: this.urlResponse
+      })
+        .as('getCapabilities');
+        
+      MapLayers
+        .openManagementModal()
+        .fillFormCorrectly(label, geoserverUrl)
```

Comment:
Wouldn't be better to call this function `fillForm`?

Reply:
In code are two function fillForm and fillFormCorrectly. It cant be in one method, because in the test correctly trail, cypress wait for request, in the test where errors are checked, wait for request it isnt possible. In WMS form I have to first fill field URL and wait for layer and next fill name, because If I make this inversely, layers from WMS are not appears without click on the modal, anywhere. So I think it is the simplier way to make two method, instead created other method for waiting request, and create an additional click on the modal.

Reply:
In such case you should consider a refactor of those 2 functions into one.

```

interface ILayer {
label: string;
url: string;
}
fillForm(form: Layer, waitForResponse = true) {
 MapLayers
      .pageObject
      .getAddBtn()
      .click();

      MapLayers
      .pageObject
      .getUrlInput()
      .type(url)

if (waitForREsponse) {cy.wait('@getCapabilities');}
     
     
      MapLayers
      .pageObject
      .getLabelInput()
      .type(label);

    return MapLayers;
}

```

After the refactoring (merge of functionalities) you could still run fillForm({...}), as also fillForm({...}, false), with little work effort.